### PR TITLE
Make sure stats are finite and rounded

### PR DESCRIPTION
**Pull request by arhimondr**
PlanNodeStatsEstimate, SymbolStatsEstimate should check that outputRowCount,
distinctValuesCount are finite and rounded. Also averageRowSize must be finite.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 1]
```diff
@@ -313,7 +314,7 @@ private PlanNodeStatsEstimate calculateJoinComplementStats(
 
         // TODO: use range methods when they have defined (and consistent) semantics
         double leftNDV = leftColumnStats.getDistinctValuesCount();
-        double matchingRightNDV = rightColumnStats.getDistinctValuesCount() * unmatchedJoinComplementNdvsCoefficient;
+        double matchingRightNDV = roundToDouble(rightColumnStats.getDistinctValuesCount() * unmatchedJoinComplementNdvsCoefficient);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: copying from https://github.com/prestodb/presto/pull/11856#discussion_r232169603

Why would you want to round NDV estimate?
1.5 is a perfect average expected value. Rounding it to 2 introduces 33% error.

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: i see your response now https://github.com/prestodb/presto/pull/11856#discussion_r232493865 (i was able to see my question but not your response on GH)

> There are no 1.5 distinct values in real world. There are either 2 or 1. Same as for the rows count. It makes it way easier to reason about the cost model when integer values are integers. In real world we cannot have 1.5 rows returned.
> 
> I had to introduce that round to make leftNDV <= matchingRightNDV check work in some of the corner cases covered by the unit tests (as value returned by SymbolStatsEstimate#getDistinctValuesCount is always rounded now).


I do not agree. 
I know the final result will be either 1 NDV or 2 (or some other integer).
However, we are not trying to "guess" the final result (in guessing, betting on a possible answer would make a lot of sense). Instead, we are trying to approximate the final result, and here, approximating with something that is not a whole integer is perfectly valid.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: I understand the expected value concept. I'm just not sure if the increased accuracy worth of always keeping in mind that NDV and Row count values could be between 0 and 1.

**Comment by arhimondr** [comment]
@sopel39 @findepi It looks like you feel strong about keeping the fractional estimates. I'm going to close this PR. In all the places where we have explicit `== 0` checks before division i'm going to check for `< 1`. Also we should probably consider checking if the value is rather close to `1` then to `0`. And if it is closer to `1` - do the rounding to 1 before the division.

**Comment by arhimondr** [comment]
@

**Pull request closed without merging**